### DISCUSSION TITLE
Need help with background_task

### DISCUSSION THREAD

**Discussion by rb1811** [fastapi/fastapi#3061] [Questions]
Hi I have read the official documentation of background_task. But one thing which I don't like about that is how we have to initialize the background_task at the root function and then carry it down all the way till that function where we have to schedule the task. I have written a code which doesn't work (I don't see "Hello" getting printed) and there is no error also. Can someone tell me if I can do something like this at all with fastapi? Any help will be highly appreciated

```
from fastapi import BackgroundTasks, FastAPI
import time

app = FastAPI()
def write_notification():
    time.sleep(5)
    print("Hello")

def funcC(param):
    time.sleep(5)
    print(param)
    bt = BackgroundTasks()
    bt.add_task(write_notification)


def funcA():
    bt = BackgroundTasks()
    bt.add_task(funcC, "World")
    return

def funcB():
    funcA()
    print("funcB")

@app.get("/")
async def send_notification():
    funcB()
    return {"message": "Notification sent in the background"}       
 
 Request
$curl http://127.0.0.1:8000

 Response
funcB
INFO:     127.0.0.1:40456 - "GET / HTTP/1.1" 200 OK
```

    
 I don't want declare background_task: BackgrounTasks in send_notification() as params then take it all way till funcA(). I want declare it directly inside funcA() is it possible in some way? Using Python-inject or basic Python somehow? 
 
 Thanks in advance.

**Comment by sssyy10677** [comment]
hi there, you should code like this way
```
app = FastAPI()
def write_notification():
    time.sleep(5)
    print("Hello")

def funcC(param, back):
    time.sleep(5)
    print(param)
    back.add_task(write_notification)


def funcA(back):
    back.add_task(funcC, "World", back)
    return

def funcB(back):
    funcA(back)
    print("funcB")

@app.get("/")
async def send_notification(back: BackgroundTasks):
    funcB(back)
    return {"message": "Notification sent in the background"}

```

**Reply by rb1811** [reply]
Yes I know this. I don't want to do this because funcB has to unnecessarily receive the "back" param when its not even scheduling the task. When the number of hops for API becomes large and you have to schedule the task at the very end of the chain all the intermediate functions need to accept and forward the "back" param which feels redundant.

**Comment by tiangolo** [comment]
If you're still having problems, it would probably be better to ask in a GitHub issue, following the template that will guide you to structure the question in a way that makes it easier to understand your intention and need, to be able to help you better. And also providing a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) that I can copy-paste to replicate it.

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Comment by extreme4all** [comment]
@rb1811
When this endpoint is called, the `send_notification()` function calls the `funcB()` function, which in turn creates a background task using the `asyncio.create_task()` function. This background task runs the `write_notification()` function, which performs some **asynchronous** operations and then prints a message to the console.

The `create_task()` function is used to create a new task that will be run in the asyncio event loop. This task is scheduled to run concurrently with other tasks in the event loop, allowing it to be executed in the background while the `send_notification()` function continues to execute. This allows the `write_notification()` function to be run as a background task, without blocking the execution of the `send_notification()` function.
```py
from fastapi import FastAPI

import uvicorn
import asyncio

app = FastAPI()

async def write_notification():
    print("Writing notification")
    await asyncio.sleep(5)
    print("Notification written")

async def funcB():
    asyncio.create_task(write_notification())
    print("funcB")

@app.get("/")
async def send_notification():
    await funcB()
    return {"message": "Notification sent in the background"} 

if __name__ == "__main__":
    uvicorn.run("background_task:app", host="localhost", port=5000, workers=1, reload=True)
```
